File: server/internal/guide/art.go
package guide

import "strconv"

// ArtLayout chooses how a picture can be shown.
// A full-bleed hero needs a landscape image at least 1280 px wide,
// and the slot must stay within 1.25× that width. Anything else is composed:
// the picture stays near its real size over a blurred copy.
func ArtLayout(width, height, slot int) string {
	if width >= 1280 && height > 0 && width > height && slot > 0 && slot*4 <= width*5 {
		return "bleed"
	}
	return "composed"
}

// DisplayEdge is the longest side to draw. Zero means the picture's own size,
// which is what we use until a probe has measured it.
func DisplayEdge(native, slot int) int {
	if native <= 0 || slot <= 0 {
		return 0
	}
	limit := native + native/4
	if slot < limit {
		return slot
	}
	return limit
}

// PickIcon prefers the largest landscape icon, then the largest icon, then the first usable URL.
func PickIcon(icons []iconEl) (src string, width, height int) {
	bestScore := -1
	for _, icon := range icons {
		url := cleanIcon(icon.Src)
		if url == "" {
			continue
		}
		w, _ := strconv.Atoi(icon.Width)
		h, _ := strconv.Atoi(icon.Height)
		score := 0
		if w > 0 {
			score = w
		}
		if w > 0 && h > 0 && w > h {
			score += 1_000_000
		}
		if src == "" || score > bestScore {
			src, width, height = url, w, h
			bestScore = score
		}
	}
	return src, width, height
}

File: web/src/lib/art.ts
/** Mirrors guide.ArtLayout. Full-bleed only for landscape art at least 1280 px wide, within 1.25×. */
export function artLayout(width: number, height: number, slot: number): "bleed" | "composed" {
  if (width >= 1280 && height > 0 && width > height && slot > 0 && slot * 4 <= width * 5) return "bleed";
  return "composed";
}

/** Longest side to draw. Zero means the picture's own pixel size. */
export function displayEdge(native: number, slot: number): number {
  if (native <= 0 || slot <= 0) return 0;
  const limit = native + Math.floor(native / 4);
  return slot < limit ? slot : limit;
}

/** CSS pixels for one side, staying within 1.25× the picture. Zero when the size is unknown. */
export function cappedCss(native: number, slotCss: number, scale = 1): number {
  if (native <= 0 || slotCss <= 0) return 0;
  const px = Math.max(scale, 1);
  const edge = displayEdge(native, Math.round(slotCss * px));
  if (edge <= 0) return 0;
  return edge / px;
}
